### ISSUE TITLE
Getting conflict when trying to modify existing key on the database

### ISSUE THREAD

**Issue by Zazzzles** [yanick/redux-persist-pouchdb#4]
Redux-persis is setup according to the docs as well as redux-persist-pouchdb,
but as soon as an action modifies state redux-persist tries to set it to the DB I get a 
conflict error.

Has anyone encountered this before? Surely old state should be overwritten in the DB when new state is set?

**Comment by williamleathers** [comment]
I'm getting the same issue:
`Error storing data CustomPouchError {status: 409, name: "conflict", message: "Document update conflict", error: true, id: "persist:myKey", …}`

**Comment by Zazzzles** [comment]
@williamleathers  I suspect that `_rev` is not being handled properly between getting and setting to the DB but I'm hoping @yanick could weigh in on this.

**Comment by williamleathers** [comment]
I'm currently trying to roll pouchdb back to v7.0.0 in case redux-persist-pouchdb only worked on that version.

**Comment by Zazzzles** [comment]
@williamleathers I switched over to `redux-pouchdb` and it seems to be working fine now
https://github.com/vicentedealencar/redux-pouchdb

**Comment by williamleathers** [comment]
Thanks @Zazzzles, I'll give that a try :)

**Comment by Zazzzles** [comment]
@williamleathers  Scratch that, `redux-pouchdb` also doesn't work. Lemme know if you find a solution :)

**Comment by williamleathers** [comment]
@Zazzzles , I got it updating the document in indexedDB, which looked promising, but then I refreshed the page and it entered an infinite loop of updating the doc. Also I need something that I can hook-up to AWS AppSync and I couldn't see how I could use this for that purpose, so also continuing to search...

**Comment by yanick** [comment]
I'll try to have a look at it today. It's been a while since I used this package and my memories are a little dusty. :-)

**Comment by Zazzzles** [comment]
@yanick Thanks a lot! Much appreciated 😄

**Comment by yanick** [comment]
First step is to admit you have a problem: I altered the tests in the branch 'key-mismatch' such that it shows the problem we're having here.

**Comment by williamleathers** [comment]
@yanick Thanks for looking in to this.

**Comment by yanick** [comment]
I have a branch (in the PR above ^^^) that should fix the issue.

The tricky part here is that we must keep the latest _rev of the document somewhere for the next update to pouchdb not to trigger a document conflict. The _persist field that redux-persist uses would have been good for it, but redux-persist makes it hard to use it, so I added another field just for this package.

Can you peeps give a whirl to that branch and see if it works for you? Note that for everything to work, you now have to add

```
storage.store = store;
```

once the redux store is created.

**Comment by yanick** [comment]
Hmmm.... I think I can simplify things a little bit by keeping the _rev values in the PouchDBStorage class itself. I have to play with this. 

TL;DR: the PR should work as-is, but if my idea works, the interface might change to something simpler soon.

**Comment by yanick** [comment]
Seems that my idea indeed works. I have updated the PR for the simplifed way that doesn't' involve playing with the store. :tada:

**Comment by Zazzzles** [comment]
@yanick Playing around with a similar implementation on my side, I also faced the issue of where to put _rev values. Ended up throwing them into localStorage which is handy since it is available for the app I'm working on. This, of course, won't work for apps that don't run in browser so then maybe putting _rev's in PouchDBStorage class itself isn't a bad idea

**Comment by williamleathers** [comment]
@yanick, thanks for working on this. However I think I must be missing something obvious. I'm trying to use the package as a custom storage engine for AWS's AppSync (aws-appsync v3.0.2), but I'm getting a `Uncaught TypeError: storage.getAllKeys is not a function` error message as if app sync is expecting a getAllKeys method to be available on it. Was this designed to work with AppSync and if so do you have any suggestions as to what I might be missing?

**Comment by yanick** [comment]
@williamleathers What you missed wasn't obvious at all. :-) While the main redux-persist docs says that the storage packages need to implement getItem / setItem / removeItem, it seems that `getAllKeys` is a method that is not a must for basic functionality, but that is expected by some upstream packages. From what I can see from the code of redux-persist itself, it might have been a part of the interface a few versions ago.

Anyway, long story short: it's easy to do, so I'll add the method today.

**Comment by yanick** [comment]
Version 0.2.1 on its way to npm!
